File: csharp/bottle-song/BottleSong.cs
using System.Collections.Generic;
using System.Text;

public static class BottleSong
{

    static Dictionary<int, string> numberLabels = new()
    {
        [0] = "Zero", [1] = "One", [2] = "Two", [3] = "Three", [4] = "Four",
        [5] = "Five", [6] = "Six", [7] = "Seven", [8] = "Eight", [9] = "Nine",
        [10] = "Ten", [11] = "Eleven", [12] = "Twelve", [13] = "Thirteen", [14] = "Fourteen",
        [15] = "Fifteen", [16] = "Sixteen", [17] = "Seventeen", [18] = "Eighteen", [19] = "Nineteen",
        [20] = "Twenty", [30] = "Thirty", [40] = "Forty", [50] = "Fifty", [60] = "Sixty",
        [70] = "Seventy", [80] = "Eighty", [90] = "Nintey", [100] = "Hundred"
    };

    public static IEnumerable<string> Recite(int startBottles, int takeDown)
    {
        for (int i = 0; i < takeDown; i++)
        {
            if (i > 0)
                yield return "";
            
            int bottleNum = startBottles - i;
            string bottle = bottleNum > 1 ? "bottles" : "bottle";
            string bottleNext;
            if (bottleNum - 1 == 0)
                bottleNext = "no green bottles";
            else if (bottleNum - 1 == 1)
                bottleNext = "one green bottle";
            else
                bottleNext = NumberToString(bottleNum - 1).ToLowerInvariant() + " green bottles";

            string firstLine = $"{NumberToString(bottleNum)} green {bottle} hanging on the wall,";
            yield return firstLine;
            yield return firstLine;
            yield return "And if one green bottle should accidentally fall,";
            yield return $"There'll be {bottleNext} hanging on the wall.";

        }
    }

    public static string NumberToString(int number)
    {
        string ret = "";
        int working = 0;
        int remainder = number;
        if (number >= 1000 || number < 0)
            throw new ArgumentException("Out of bounds, only 0-999 are supported.");
        if (number >= 100)
        {
            remainder = number % 100;
            working = (number - remainder) / 100;
            if (ret.Length > 0) ret = ret + " ";
            ret = ret + NumberToString(working) + " " + numberLabels[100];
            number = remainder;
        }

        if (number <= 20)
        {
            if (ret.Length > 0) ret = ret + " ";
            ret = ret + numberLabels[number];
        }
        else
        {
            remainder = number % 10;
            working = number - remainder;
            if (ret.Length > 0) ret = ret + " ";
            ret = ret + numberLabels[working] + " " + numberLabels[remainder];
        }

        return ret;
    }
}

File: javascript/bottle-song/bottle-song.js
//
// This is only a SKELETON file for the 'Bottle Song' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

const numOneToNineteen = ['zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 
  'ten', 'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen'
]

const numTens = ['zero', 'ten', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety', 'one hundred']

/**
 * Convert a number 0-100 to text. Only integer values are supported (no 1.25)
 * @param {Number} num - the value to convert to a text representation
 * @throws {Error} throws an error if the number is not in the range of 0 to 100. 
 */
const numToText = (num) => {
  if (num < 0 || num > 100)
    throw new Error("Index out of Range.")

  if (num < 20)
    return numOneToNineteen[num]

  let ones = num % 10
  let tens = (num - ones) / 10

  if (ones === 0)
    return numTens[tens]  
  return numTens[tens] + " " + numOneToNineteen[ones]
}

const capitalize = (text) => {
  if (text.length === 0)
    return ''
  return `${text.charAt(0).toUpperCase()}${text.slice(1)}`;
}

/**
 * 
 * @param {Number} initialBottlesCount - Number of bottles to start with
 * @param {Number} takeDownCount - How many verses to write.
 */
export const recite = (initialBottlesCount, takeDownCount) => {
  /*
  `Three green bottles hanging on the wall,`,
  `Three green bottles hanging on the wall,`,
  `And if one green bottle should accidentally fall,`,
  `There'll be two green bottles hanging on the wall.`,
  */
  let currentCount = initialBottlesCount;
  let ret = []
  for(let i = 0; i < takeDownCount && currentCount > 0; i++)
  {
    let nextCount = currentCount - 1

    ret.push(`${capitalize(numToText(currentCount))} green ${currentCount === 1 ? 'bottle' : 'bottles'} hanging on the wall,`)
    ret.push(`${capitalize(numToText(currentCount))} green ${currentCount === 1 ? 'bottle' : 'bottles'} hanging on the wall,`)
    ret.push('And if one green bottle should accidentally fall,')
    ret.push(`There'll be ${nextCount === 0 ? 'no' : numToText(nextCount)} green ${nextCount === 1 ? 'bottle' : 'bottles'} hanging on the wall.`)
    ret.push('')
    currentCount = nextCount
  }
  ret.pop() // remove last blank line
  return ret
};
